week-2-wide-perspective/w02e04-dungeons-and-the-princess/src/models/Error.ts

```diff
@@ -0,0 +1,28 @@
+import { Castle } from './Castle';
+import { Dungeon } from './Dungeon';
+import { Underground } from './Underground';
+import { Person } from './Person';
+
+class PersonNotAllowedError extends Error {
+  constructor({
+    place,
+    person
+  }: {
+    place: Castle | Dungeon | Underground;
+    person: Person;
+  }) {
+    super();
+    this.message = `${
+      person.title
+    } is not allowed to be in the ${place.toString()}`;
+  }
+}
+
+class InvalidSaviorError extends Error {
+  constructor(person: Person) {
+    super();
+    this.message = `${person.title} cannot attempt to save princess`;
+  }
+}
+
+export { PersonNotAllowedError, InvalidSaviorError };
```

Comment:
Bardzo ładne podejście - o to chodziło, o dedykowane `error`y.

Generalnie zadanie jest rozpykane, 
Twoje błędy obsługują miejsce z którego się wywodzą. Jednakże - nawet zbyt dosłownie :)

CO do `place` to podczepiasz sobie instancję miejsca - ale nie trzymasz jej w żadnym dodatkowym polu.
Np powinno być:
```typescript
this.place = place
```

tak aby to zapisać w Error.

Dodatkowo jednak. Zmieniłbym delikatnie ten sposób dostarczania informacji "skąd jest błąd".

Zobacz.

Gdybyś wprowadził sobie odpowiednie klasy dla miejsc z których pochodzą errory:
`CastleError` / `UndergroundError` / `DungeonError` 

to sama klasa determinowała by pochodzenie:

np.
```typescript
if(error instanceof CastleError) {
  console.log('Error from the Castle`)
  console.log(error.message)
}
```

Unikamy przesyłania instancji miejsca (nie jest nam to potrzebne, bo w zależności od mniejsaca używamy odpowiedniej klasy errora do rzucania)
Unikamy enuma
Unikamy boilerplate (choć dostarczamy inny związany z klasami błędów

Wtedy dostajesz symulację rodem z `back-endu` z którą spotkasz się jeszcze nie raz.
Właśnie na back-endzie.

Rzucany jest error, a `globalny error handler` - sprawdza skąd on pochodzi, czy `FileSystem` czy `DatabaseError` czy `FileUploadError` czy `ControllerError` czy ... itp itd.

oto challenge dla Ciebie, wprowadzić sobie właśnie podział na klasy Errorów które będziesz używał w zależności od miejsca :)